### ISSUE TITLE
Spike: top level partition context

### ISSUE THREAD

**Issue by npepinpe** [camunda-cloud/zeebe#7637]
**Description**

The goal of this spike to figure out the steps needed to merge the "RaftPartition" and "ZeebePartition" context into a single, top level "Partition" context. The focus is primarily on startup/shutdown and components inter dependencies of the immediate components (e.g. Journal, SnapshotStore, RaftLog, RaftServer, etc.) and not refactoring those components themselves (which will be a future step).

**Comment by npepinpe** [comment]
Notes from the spike:

- Using this huge mutable context is error prone and awkward: there are implicit dependencies between the steps (i.e. StepB depends on component A in the context, which is set by StepA, so StepA must be run before), the context ends up being gigantic. I would like to investigate after we migrate to the `StartupProcess` if maybe there isn't a better alternative, something more like an aggregating pipeline but where the dependencies are explicit. However, this is out of scope for now, since we already have the context (meaning we won't make things worse), and introducing the process will improve other aspects, so it's still worth doing that first.
- I would like to introduce higher level components which provide a simple API for a [deep module](https://medium.com/@nakabonne/depth-of-module-f62dac3c2fdb) (note: I didn't read the article, but I did read the book which it summarizes, and this is the definition I'm referring to :smile:). So for example, I would like a Raft component (which would be the first draft to later have a Raft module) which is the single entry point for non-Raft components to interact with Raft. I think we saw good wins in the Journal when we refactored it like this by cutting out a lot of interfaces and reducing the API surface when it comes to external dependents, and I would like to apply the same here. It would also simplify the partition bootstrap/transition by encapsulating the more granular bootstrap/transition behaviors, such that the Partition's startup can go from `[snapshot store, meta store, raft log, raft server, log stream, state controller, engine, exporter, snapshot director]` to something like `[snapshot store, raft, log stream, state controller, engine, exporter, snapshot director]`
- Eventually it will make sense to allow some steps to be executed in parallel, but I think this adds extra complexity and we don't need this now. We can make this possible later on, as it's much easier to reason about sequential steps.
- Here's the startup dependency break down as I see it, from bottom to top; arrows go from dependent to dependency, e.g. B depends on A is annotated `A <- B`

![Partition components](https://user-images.githubusercontent.com/43373/131650254-2711abd5-78ec-4d1e-99b0-ad5e608a81eb.png)

`Raft` here can be an encapsulation of the whole thing - it doesn't need to be present in the first iteration, but could be later on. The `Journal` depends on the `MetaStore` for the `lastWrittenIndex`, and since each can fail to start, I split them up (since the `MetaStore` step can fail due to I/O). Some notes on the `MetaStore`: set up is done in the constructor and should be extracted. It's also coupled to the `RaftStorage`, but we can simplify this. The `Storage` box is also maybe unnecessary, we can discuss it. What it does is: ensure the directory exists, and lock/reads the lock file, and verifies we can use this directory. This is also an operation that can fail, however it's only a start up step (there is no shutdown equivalent), so maybe it's not a step but just...something we do? Also ensuring the directory exists should probably be done at the partition level before the snapshot store to be honest, so then it's just handling the lock file. I kept the `RaftLog` in the `RaftServer` for now, since it's just a wrapper around the journal and nothing else. Note also that this is not a module dependency graph, but just a startup dependency - so that's why you don't see `ZeebeDb` for example, since the startup step is really the `StateController` step, which owns the `ZeebeDb` instance. I also initially had the snapshot director as part of the state controller, but I'm not so sure now since the snapshot director needs to know about the last processed position from the engine, so it may be necessary to split them up.
